### DISCUSSION TITLE
Apply the constitutive model considering the effect of grain size, dislocation interaction, second phase particles and temperature to polycrystal and solve it with VPSC in Moose

### DISCUSSION THREAD

**Discussion by xchengood** [idaholab/moose#19418] [Q&A Modules: Solid mechanics] [answered]
Hi Moose experts or users,

A constitutive model taking the effect of grain size, dislocation interaction (thermal glide and climb mechanism), second phase strengthening, and temperature is good to study materials' mechanical properties and creep behaviours at different temperatures. There are many such models considering one or more factors proposed. For example, a constitutive model for only the glide and evolution of forest dislocations within a Hexagonal Close-Packed (HCP) crystal lattice (https://mooseframework.inl.gov/source/materials/crystal_plasticity/CrystalPlasticityHCPDislocationSlipBeyerleinUpdate.html). And Moose team already built this class well. 

The old discussion post is made here (https://github.com/idaholab/moose/discussions/19199). Based on that, I have the following two questions.

1. As mentioned, 'the VPSC is a kind of solver and a constitutive model', is there a VPSC solver in Moose? If so, how to use it and are there any examples for reference? If not, How can I apply the HCP model to polycrystal, not only for a single crystal?

2. I also want to study how the stress-strain curve change with temperature. I add the the following block to the [AuxKernels]
[temperature]
type = FunctionAux
variable = temperature
function = '300+400*t' # temperature increases at a constant rate
execute_on = timestep_begin
[]
Is my method correct? Basically, I got the stress-strain curve at 300K and 1600K. The yield strength shows a significant decrease, but Young's modulus does not show any change. Two curves have the same slope as shown below. This does not make sense. If my method is not correct, could you share any ideas with me? Basically, Young's modulus should decrease with the increase of temperature.
<img width="123" alt="Screen Shot 2021-11-16 at 4 07 58 PM" src="https://user-images.githubusercontent.com/74226455/142074055-646e2faa-63d8-43d3-b771-c3453bcf4353.png">

Thank you. @sapitts @ngrilli

**Comment by ngrilli** [comment, marked as answer]
Dear @xchengood, 

I don't think there is a VPSC solver, just use the crystal plasticity formulation in MOOSE.

You can find an example polycrystal simulation and instruction about how to create a polycrystal using Neper here:
https://github.com/ngrilli/c_pfor_am/tree/main/test/tests/neper

The plasticity material model will not change the elastic modulus with temperature.
The elastic modulus is defined in the material object ComputeElasticityTensorConstantRotationCP
which is not temperature dependent.

I have made a new material class that you can use for the temperature dependence of the elastic modulus:
https://github.com/ngrilli/c_pfor_am/blob/main/src/materials/ComputeElasticityTensorCPGrain.C

There is a test case here:
https://github.com/ngrilli/c_pfor_am/blob/main/test/tests/TempDepElasticConstants/TempDepElasticConst.i

The class should be compatible with the existing CP models in MOOSE
but you need to check for that.

Best Regards,
Nicolò
